File: adk-guide/blog_05_adk_for_typescript/02_custom_tool_agent.ts
/**
 * Blog 05 — ADK for TypeScript (December 2025)
 * Example 2: Agent with Custom TypeScript Tools
 *
 * Source: https://developers.googleblog.com/introducing-agent-development-kit-for-typescript-build-ai-agents-with-the-power-of-a-code-first-approach/
 *
 * Custom TypeScript functions can be registered as ADK tools using FunctionTool.
 * The type annotations and JSDoc comments help ADK generate the correct schema
 * for the LLM to understand when and how to call each function.
 *
 * Run:
 *   export GOOGLE_API_KEY="your-api-key"
 *   npx ts-node 02_custom_tool_agent.ts
 */

import { LlmAgent, FunctionTool, InMemorySessionService, Runner } from '@google/adk';

// ── Tool types ────────────────────────────────────────────────────────────────
interface WeatherResult {
  city: string;
  temperature: string;
  condition: string;
  humidity: string;
  feelsLike: string;
}

interface ConversionResult {
  original: number;
  originalUnit: string;
  converted: number;
  convertedUnit: string;
}

// ── Tool 1: Weather lookup ────────────────────────────────────────────────────
/**
 * Retrieves the current weather for a specified city.
 * @param city - The city name to look up weather for (e.g., "Tokyo", "London")
 * @returns Current weather data including temperature, condition, and humidity
 */
function getWeather(city: string): WeatherResult {
  const weatherData: Record<string, WeatherResult> = {
    london: {
      city: 'London',
      temperature: '15°C',
      condition: 'Cloudy',
      humidity: '78%',
      feelsLike: '13°C',
    },
    tokyo: {
      city: 'Tokyo',
      temperature: '22°C',
      condition: 'Sunny',
      humidity: '55%',
      feelsLike: '24°C',
    },
    'new york': {
      city: 'New York',
      temperature: '18°C',
      condition: 'Partly Cloudy',
      humidity: '62%',
      feelsLike: '17°C',
    },
    sydney: {
      city: 'Sydney',
      temperature: '25°C',
      condition: 'Clear',
      humidity: '45%',
      feelsLike: '27°C',
    },
  };

  return (
    weatherData[city.toLowerCase()] ?? {
      city,
      temperature: 'N/A',
      condition: 'Unknown',
      humidity: 'N/A',
      feelsLike: 'N/A',
    }
  );
}

// ── Tool 2: Currency conversion ───────────────────────────────────────────────
/**
 * Converts an amount from one currency to another using approximate rates.
 * @param amount - The numeric amount to convert
 * @param fromCurrency - The source currency code (e.g., "USD", "EUR", "GBP")
 * @param toCurrency - The target currency code (e.g., "USD", "EUR", "JPY")
 * @returns The converted amount with both original and converted values
 */
function convertCurrency(
  amount: number,
  fromCurrency: string,
  toCurrency: string
): ConversionResult {
  // Approximate exchange rates relative to USD
  const ratesUSD: Record<string, number> = {
    USD: 1.0,
    EUR: 0.92,
    GBP: 0.79,
    JPY: 149.5,
    AUD: 1.52,
    CAD: 1.36,
    INR: 83.2,
    CHF: 0.9,
  };

  const fromRate = ratesUSD[fromCurrency.toUpperCase()] ?? 1;
  const toRate = ratesUSD[toCurrency.toUpperCase()] ?? 1;

  // Convert: source → USD → target
  const inUSD = amount / fromRate;
  const converted = inUSD * toRate;

  return {
    original: amount,
    originalUnit: fromCurrency.toUpperCase(),
    converted: Math.round(converted * 100) / 100,
    convertedUnit: toCurrency.toUpperCase(),
  };
}

// ── Tool 3: Text statistics ───────────────────────────────────────────────────
/**
 * Calculates statistics for a given text string.
 * @param text - The text to analyse
 * @returns Statistics including word count, character count, and reading time
 */
function analyseText(text: string): {
  wordCount: number;
  charCount: number;
  sentenceCount: number;
  avgWordLength: number;
  estimatedReadingTimeSeconds: number;
} {
  const words = text.trim().split(/\s+/).filter(Boolean);
  const sentences = text.split(/[.!?]+/).filter(Boolean);
  const avgWordLen =
    words.reduce((sum, w) => sum + w.length, 0) / (words.length || 1);

  return {
    wordCount: words.length,
    charCount: text.length,
    sentenceCount: sentences.length,
    avgWordLength: Math.round(avgWordLen * 10) / 10,
    estimatedReadingTimeSeconds: Math.ceil((words.length / 200) * 60), // avg 200 wpm
  };
}

// ── Define the agent with multiple tools ──────────────────────────────────────
const utilityAgent = new LlmAgent({
  name: 'utility_assistant',
  model: 'gemini-2.5-flash',
  description:
    'A versatile assistant with weather lookup, currency conversion, and text analysis.',
  instruction:
    'You are a helpful utility assistant. Use your tools to: ' +
    'check weather with getWeather, convert currencies with convertCurrency, ' +
    'and analyse text with analyseText. Always present results clearly.',
  tools: [
    new FunctionTool(getWeather),
    new FunctionTool(convertCurrency),
    new FunctionTool(analyseText),
  ],
});

// ── Run queries ───────────────────────────────────────────────────────────────
async function main(): Promise<void> {
  const sessionService = new InMemorySessionService();
  const session = await sessionService.createSession({
    appName: 'utility_app',
    userId: 'user_01',
  });

  const runner = new Runner({
    agent: utilityAgent,
    appName: 'utility_app',
    sessionService,
  });

  const queries = [
    "What's the weather in Tokyo?",
    'Convert 500 USD to Japanese Yen.',
    'Analyse this text: "Google ADK makes it incredibly easy to build powerful multi-agent AI systems in TypeScript."',
  ];

  for (const query of queries) {
    console.log(`User: ${query}`);

    const events = runner.runAsync({
      userId: 'user_01',
      sessionId: session.id,
      newMessage: { role: 'user', parts: [{ text: query }] },
    });

    for await (const event of events) {
      if (event.isFinalResponse()) {
        console.log(`Agent: ${event.content?.parts?.[0]?.text ?? ''}\n`);
      }
    }
  }
}

main().catch(console.error);

File: adk-guide/blog_01_initial_launch_multi_agent/02_custom_tool_agent.py
"""
Blog 01 — Google ADK Initial Launch (April 2025)
Example 2: Agent with a Custom Python Tool

Source: https://developers.googleblog.com/en/agent-development-kit-easy-to-build-multi-agent-applications/

ADK treats any Python function as a tool. The docstring is used by the LLM
to understand when and how to call the function.
"""

import asyncio
from google.adk.agents import Agent
from google.adk.runners import Runner
from google.adk.sessions import InMemorySessionService
from google.genai import types


# ── Custom tool: weather lookup (mocked) ─────────────────────────────────────
def get_weather(city: str) -> dict:
    """
    Retrieves the current weather for a given city.

    Args:
        city: The name of the city to get weather for (e.g., "London", "Tokyo").

    Returns:
        A dictionary containing 'temperature', 'condition', and 'humidity'.
    """
    # In production, call a real weather API here.
    mock_data = {
        "london":  {"temperature": "15°C", "condition": "Cloudy",  "humidity": "78%"},
        "tokyo":   {"temperature": "22°C", "condition": "Sunny",   "humidity": "55%"},
        "new york":{"temperature": "18°C", "condition": "Partly Cloudy", "humidity": "62%"},
        "sydney":  {"temperature": "25°C", "condition": "Clear",   "humidity": "45%"},
    }
    return mock_data.get(city.lower(), {
        "temperature": "N/A",
        "condition":   "Unknown",
        "humidity":    "N/A",
    })


# ── Custom tool: unit conversion ──────────────────────────────────────────────
def convert_temperature(value: float, from_unit: str, to_unit: str) -> dict:
    """
    Converts a temperature value between Celsius, Fahrenheit, and Kelvin.

    Args:
        value: The numeric temperature value to convert.
        from_unit: The source unit — 'celsius', 'fahrenheit', or 'kelvin'.
        to_unit: The target unit — 'celsius', 'fahrenheit', or 'kelvin'.

    Returns:
        A dictionary with 'result' (float) and 'unit' (str).
    """
    # Normalise to Celsius first
    celsius_map = {
        "celsius":    lambda v: v,
        "fahrenheit": lambda v: (v - 32) * 5 / 9,
        "kelvin":     lambda v: v - 273.15,
    }
    to_map = {
        "celsius":    lambda c: c,
        "fahrenheit": lambda c: c * 9 / 5 + 32,
        "kelvin":     lambda c: c + 273.15,
    }
    celsius = celsius_map[from_unit.lower()](value)
    result  = to_map[to_unit.lower()](celsius)
    return {"result": round(result, 2), "unit": to_unit}


# ── Define the agent ──────────────────────────────────────────────────────────
weather_agent = Agent(
    name="weather_assistant",
    model="gemini-2.5-flash",
    instruction=(
        "You are a weather assistant. Use the get_weather tool to fetch current "
        "conditions for any city. If asked about temperatures in different units, "
        "use the convert_temperature tool."
    ),
    description="An agent that provides weather information and temperature conversions.",
    tools=[get_weather, convert_temperature],
)


# ── Run the agent ─────────────────────────────────────────────────────────────
async def main():
    session_service = InMemorySessionService()
    session = await session_service.create_session(
        app_name="weather_app", user_id="user_01"
    )
    runner = Runner(
        agent=weather_agent,
        app_name="weather_app",
        session_service=session_service,
    )

    queries = [
        "What is the weather like in Tokyo right now?",
        "Convert 22 degrees Celsius to Fahrenheit.",
    ]

    for query in queries:
        print(f"User: {query}")
        async for event in runner.run_async(
            user_id="user_01",
            session_id=session.id,
            new_message=types.Content(
                role="user", parts=[types.Part(text=query)]
            ),
        ):
            if event.is_final_response():
                print(f"Agent: {event.content.parts[0].text}\n")


if __name__ == "__main__":
    asyncio.run(main())
